### ISSUE TITLE
No checkpoints in games generated via Custom (Advanced) Difficulty

### ISSUE THREAD

**Issue by ghost** [Coolcord/Level-Headed#10]
As the title says, games generated with a custom difficulty seem to lack checkpoints. I'll have to try the preset difficulties later to see if those are affected as well. Again, this applies to both 0.3-RC2 and 0.3-RC3.

Traditionally, checkpoints in SMB1 are suppose to be present in all levels except castles and World 8 (*or whatever the final world becomes in the randomizer*)

Enclosed is the level file so you can try and figure out what happened.
[Random 18-07-27-21-44-49-278.lvls.txt](https://github.com/Coolcord/Level-Headed/files/2237648/Random.18-07-27-21-44-49-278.lvls.txt)

**Comment by Coolcord** [comment]
This isn't a bug.

Checkpoints were disabled in v0.2.0 due to game engine limitations and I have no plans to bring them back. SMB1 stores checkpoints in nibbles. These nibbles map out to levels 1-4 for each world in a linear fashion. If you have an 8 level world, checkpoints will break, as the system ends up wrapping around in this nibble allocation. You can still add checkpoints by hand if you really want by editing level scripts, but they will break if you have worlds with more than 4 levels.

In the first public release, midpoints had to be accounted for while generating a level. However, there were many factors that could cause them to break, and I ultimately didn't think they were worth my time investment, so I traded them out for 5+ level worlds.

Thanks anyway for the bug report, and I hope this makes sense.

**Comment by ghost** [comment]
I see, so if the feature is dependent of having 4 level worlds, couldn't you just have some sort of if-then branch to enable it only if it's manually set to 4 levels per world? It'd make playing with crazy difficulties less frustrating and more enjoyable to explore imo.

**Comment by Coolcord** [comment]
It's more complicated than that. An entire midpoint handler would need to be written that can handle every level type. Adding new patterns to the project can also break the midpoint handler easily. If this project is going to ever support user created patterns, it would need to be dropped.

In short, the SMB1 engine has a hard-coded spawn coordinate that is only valid every 16 blocks. 2 blocks must be clear for this to work cleanly, and both must be at ground level. In other words, if the generator is in the middle of a pattern when it passes the midpoint, problems can arise. While you could push back the midpoint in this case, what you get left with is a lot of empty space. These issues are only hitting the tip of the iceberg here too.

If you really want midpoints, try the [Water Vapor Build (v0.1)](https://github.com/Coolcord/Level-Headed/releases/tag/v0.1.0) as I tried to support midpoints back then. [Here](https://github.com/Coolcord/Level-Headed/blob/44cbcd6de3a101744e61afeb3924ed3ff9a3c6ea/SMB1/SMB1_Compliance_Generator/Midpoint_Handler.cpp) is what the code looked like for midpoints back in v0.1. After playing it for a while, you might run into bugs with the midpoints. If you hit a bug, you can easily get put in a position where Mario gets killed infinitely until you get a Game Over.

In short, adding midpoints is much, much easier said than done, and I really do think my time would be better spent elsewhere.

**Comment by ghost** [comment]
I see, thanks for explaining that to me. Sorry to bother you over it.
